bilge-impl/src/bitsize/split.rs

```diff
@@ -0,0 +1,185 @@
+use proc_macro_error::{abort_call_site, abort};
+use quote::ToTokens;
+use syn::{meta::ParseNestedMeta, Path, Item, Attribute, Meta, parse_quote};
+use crate::shared::unreachable;
+
+/// Since we want to be maximally interoperable, we need to handle attributes in a special way.
+/// We use `#[bitsize]` as a sort of scope for all attributes below it and
+/// the whole family of `-Bits` macros only works when used in that scope.
+/// 
+/// Let's visualize why this is the case, starting with some user-code:
+/// ```ignore
+/// #[bitsize(6)]
+/// #[derive(Clone, Copy, PartialEq, DebugBits, FromBits)]
+/// struct Example {
+///     field1: u2,
+///     field2: u4,
+/// }
+/// ```
+/// First, the attributes get sorted, depending on their name.
+/// Every attribute in need of field information gets resolved first,
+/// in this case `DebugBits` and `FromBits`.
+/// 
+/// Now, after resolving all `before_compression` attributes, the halfway-resolved
+/// code looks like this:
+/// ```ignore
+/// #[bilge::bitsize_internal(6)]
+/// #[derive(Clone, Copy, PartialEq)]
+/// struct Example {
+///     field1: u2,
+///     field2: u4,
+/// }
+/// ```
+/// This `#[bitsize_internal]` attribute is the one actually doing the compression and generating
+/// all the getters, setters and a constructor.
+/// 
+/// Finally, the struct ends up like this (excluding the generated impl blocks):
+/// ```ignore
+/// struct Example {
+///     value: u6,
+/// }
+/// ```
+pub struct SplitAttributes {
+    pub before_compression: Vec<Attribute>,
+    pub after_compression: Vec<Attribute>,
+}
+
+/// Split item attributes into those applied before bitfield-compression and those applied after.
+/// Also, abort on any invalid configuration.
+/// 
+/// Any derives with suffix `Bits` will be able to access field information.
+/// This way, users of `bilge` can define their own derives working on the uncompressed bitfield.
+pub fn split_item_attributes(item: &Item) -> SplitAttributes {
```

Comment:
Do we want `SplitAttributes::new(item: &Item)`? I don't know

Reply:
oh yeah, that's a good idea.